### PR TITLE
More cleaning uses of arithmetic subtyping

### PR DESCRIPTION
**Pull request by ajreynol**
Towards eliminating arithmetic subtyping.

### CODE REVIEW

File: src/theory/quantifiers/sygus/sygus_grammar_cons.cpp [hunk 1]
```diff
@@ -794,10 +794,10 @@ void CegGrammarConstructor::mkSygusDefaultGrammar(
         // make data type for positive constant integers
         sdts.push_back(SygusDatatypeGenerator(pos_int_name));
         /* Add operator 1 */
-        Trace("sygus-grammar-def") << "\t...add for 1 to Pos_Int\n";
+        Trace("sygus-grammar-def") << "\t...add for 1.0 to Pos_Int\n";
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/sygus_grammar_cons.cpp:797]: This whole block should be adapted to not refer to ints but to rationals, I think. The names and docs are all wrong.

File: src/theory/rewriter.cpp [hunk 2]
```diff
@@ -219,6 +219,10 @@ Node Rewriter::rewriteTo(theory::TheoryId theoryId,
           TheoryId newTheory = theoryOf(newNode);
           rewriteStackTop.d_node = newNode;
           rewriteStackTop.d_theoryId = newTheory;
+          Assert(
```
Comment (HanielB) [inline src/theory/rewriter.cpp:222]: What is the motivation for this assertion?

File: src/theory/arith/theory_arith_private.cpp [hunk 3]
```diff
@@ -4456,12 +4459,21 @@ bool TheoryArithPrivate::rowImplicationCanBeApplied(RowIndex ridx, bool rowUp, C
...
+                         [nm](const Rational& r) { return nm->mkConstInt(r); });
+        }
+        else
+        {
+          std::transform(
+              coeffs->begin(),
+              coeffs->end(),
+              std::back_inserter(farkasCoefficients),
+              [nm](const Rational& r) { return nm->mkConstReal(r); });
+        }
```
Comment (HanielB) [inline src/theory/arith/theory_arith_private.cpp:4476]:
```suggestion
        std::transform(coeffs->begin(),
                       coeffs->end(),
                       std::back_inserter(farkasCoefficients),
                       [nm, type](const Rational& r) {
                         return nm->mkConstRealOrInt(type, r);
                       });
```

File: src/theory/arith/nl/ext/ext_state.cpp [hunk 4]
```diff
@@ -37,9 +37,12 @@ ExtState::ExtState(InferenceManager& im, NlModel& model, Env& env)
...
   d_true = NodeManager::currentNM()->mkConst(true);
-  d_zero = NodeManager::currentNM()->mkConst(CONST_RATIONAL, Rational(0));
-  d_one = NodeManager::currentNM()->mkConst(CONST_RATIONAL, Rational(1));
-  d_neg_one = NodeManager::currentNM()->mkConst(CONST_RATIONAL, Rational(-1));
+  // Note that we use reals for these constants. They are used in some contexts
+  // for e.g. comparing whether a monomial is greater than 1, where this is
+  // done in a way that is agnostic to types.
+  d_zero = NodeManager::currentNM()->mkConstReal(Rational(0));
+  d_one = NodeManager::currentNM()->mkConstReal(Rational(1));
+  d_neg_one = NodeManager::currentNM()->mkConstReal(Rational(-1));
```
Comment (HanielB) [inline src/theory/arith/nl/ext/ext_state.cpp:45]: Maybe best to just delete these and have the places that use them to make the proper checks?

File: src/theory/arith/theory_arith_private.cpp (hunk 3 above)
Comment (ajreynol) [reply src/theory/arith/theory_arith_private.cpp:4476]: This doesn't work, I'm not sure why. It complains that `type` is not bound.

File: src/theory/arith/nl/ext/ext_state.cpp (hunk 4 above)
Comment (ajreynol) [reply src/theory/arith/nl/ext/ext_state.cpp:45]: Yeah, in fact my latest changes do delete these.  I will delete them in a followup PR.  I can undo the change to this block.

File: src/theory/rewriter.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/rewriter.cpp:222]: This is a key invariant that the rewriter should maintain: that rewriting preserves types.  Note that `isSubtypeOf` will be replaces by equality of types when subtyping is eliminated.

Comment (HanielB) [reply src/theory/rewriter.cpp:222]: Ah, OK.

File: src/theory/arith/nl/ext/ext_state.cpp (hunk 4 above)
Comment (HanielB) [reply src/theory/arith/nl/ext/ext_state.cpp]: OK

**Review by HanielB — APPROVED** [review]

**Pull request merged**